### ISSUE TITLE
Fix error handling (BaseRequest.badRequest() etc)

### ISSUE THREAD

**Issue by dmitrizagidulin** [anvilresearch/oidc-op#4]
Current behavior - returning badRequest(whatever). For example, in TokenRequest.verifyAuthorizationCode():

```js
          return request.badRequest({
            error: 'invalid_grant',
            error_description: 'Authorization code expired'
          })
```
The problem with this approach is that it doesn't break out of the promise chain, and continues onto grant(), and fails on the first line of that function:

```
  grant (request) {
    let {grantType} = request
```

(Since what's being passed into grant() is not a token request, but a badRequest() promise.

Anyways, this error handling flow needs to be refactored, and needs to rely on `throw`, to break out of the promise chain.

**Comment by dmitrizagidulin** [comment]
One other thing to refactor, in terms of OP and error handling.
All the API handler methods (authorize(), discover(), jwks(), etc) that currently take `req, res` also need to take the Node `next` argument, used for Node-style error handling.

So, authorize will look like:

```js
  authorize (req, res, next) {
    AuthenticationRequest.handle(req, res, this, next)
  }
```

Similarly, downstream, in the various promises and their error handling, they need to use Node-style error handling as well. So, all their promise `.catch(err)` statements will need to look like:

```js
  .catch(err => {
    next(err)
  })
```

**Comment by dmitrizagidulin** [comment]
Fixed in PR #24, closing.
